R/differentiate-expr.R

```diff
@@ -0,0 +1,291 @@
+    if (is.numeric(a) && is.numeric(b)) {
+      a + b
+    } else if (is.numeric(b)) {
+      plus(b, a)
+    } else if (is_zero(a)) {
+      .drop_parens(b)
+    } else if (is_call(b, "+")) {
+      plus(plus(a, b[[2]]), b[[3]])
+    } else {
+      call("+", a, b)
+    }
+  }
+  minus <- function(a, b) {
+    if (is.numeric(a) && is.numeric(b)) {
+      a - b
+    } else if (is_zero(b)) {
+      .drop_parens(a)
+    } else if (is_zero(a)) {
+      uminus(b)
+    } else {
+      call("-", a, .protect(b, c("*", "/", "^")))
+    }
+  }
+  uminus <- function(a) {
+    if (is.numeric(a)) {
+      -a
+    } else if (.is_unary_minus(a)) {
+      a[[2]]
+    } else if (is_call(a, "*")) {
+      if (.is_unary_minus(a[[2]])) {
+        times(a[[2]][[2]], a[[3]])
+      } else if (.is_unary_minus(a[[3]])) {
+        times(a[[2]], a[[3]][[2]])
+      } else {
+        times(uminus(a[[2]]), a[[3]])
+      }
+    } else if (is_call(a, "/")) {
+      maths$divide(maths$uminus(a[[2]]), a[[3]])
+    } else if (is_call(a, "-") && length(a) == 3) {
+      minus(a[[3]], a[[2]])
+    } else if (is_call(a, "(")) {
+      uminus(.drop_parens(a))
+    } else {
+      call("-", .protect(a, c("*", "/", "^")))
+    }
+  }
+  times <- function(a, b) {
+    if (is.numeric(a) && is.numeric(b)) {
+      a * b
+    } else if (is.numeric(b)) {
+      ## below here, we can assume that 'b' is a language expression
+      times(b, a)
+    } else if (is_zero(a)) {
+      0
+    } else if (is_one(a)) {
+      rewrite(b)
+    } else if (is_minus_one(a)) {
+      uminus(b)
+    } else if (is_call(a, "/")) {
+      ## we have (a2 / a3 * b -> a2 * b / a3)
+      divide(times(a[[2]], b), a[[3]])
+    } else if (is_call(b, "/")) {
+      ## we have (a * (b2 / b3)) -> (a * b2) / b3
+      divide(times(a, b[[2]]), b[[3]])
+    } else if (is_call(b, "*") && is.numeric(b[[2]])) {
+      times(times(a, b[[2]]), b[[3]])
+    } else {
+      if (.is_unary_minus(b, TRUE)) {
+        a <- uminus(a)
+        b <- uminus(b)
+      }
+      aa <- .protect(a, c("*", "unary_minus", "/", "^"))
+      bb <- .protect(b, c("*", "^"))
+      if (is_call(bb, "*")) {
+        call("*", call("*", aa, bb[[2]]), bb[[3]])
+      } else {
+        call("*", aa, bb)
+      }
+    }
+  }
+  divide <- function(a, b) {
+    if (is.numeric(a) && is.numeric(b)) {
+      a / b
+    } else if (is_one(b)) {
+      a
+    } else if (is_zero(a)) {
+      0
+    } else if (is_zero(b)) {
+      Inf
+    } else if (is_call(a, "/")) {
+      divide(a[[2]], times(a[[3]], b))
+    } else if (is_call(b, "/")) {
+      times(a, divide(b[[3]], b[[2]]))
+    } else {
+      if (.is_unary_minus(b, TRUE)) {
+        a <- uminus(a)
+        b <- uminus(b)
+      }
+      call("/", .protect(a, c("*", "unary_minus", "^")), .protect(b, "^"))
+    }
+  }
+  pow <- function(a, b) {
+    if (is.numeric(a) && is.numeric(b)) {
+      a^b
+    } else if (is_one(b)) {
+      a
+    } else if (is.numeric(b) && b == 2 && is.symbol(a)) {
```

Comment:
I know we aren't doing all the simplifications in the world but seems pretty natural to put
```suggestion
    } else if (is_one(b)) {
      a
    } else if (is_zero(b)) {
      1
    } else if (is.numeric(b) && b == 2 && is.symbol(a)) {
```